kroxylicious/src/main/java/io/kroxylicious/proxy/internal/InternalCompletionStage.java

```diff
@@ -18,226 +19,238 @@
  * to block the proxy thread loop using the CompletionStage we offer though the kroxylicious
  * filter api, or complete the underlying future unexpectedly.
  */
-public class InternalCompletionStage<T> implements CompletionStage<T> {
+class InternalCompletionStage<T> implements CompletionStage<T> {
 
     private final CompletionStage<T> completionStage;
+    private final Executor executor;
 
-    public InternalCompletionStage(CompletionStage<T> completionStage) {
+    InternalCompletionStage(CompletionStage<T> completionStage, Executor executor) {
+        Objects.requireNonNull(completionStage);
+        Objects.requireNonNull(executor);
         this.completionStage = completionStage;
+        this.executor = executor;
+    }
+
+    private <U> CompletionStage<U> wrap(CompletionStage<U> completionStage) {
+        return new InternalCompletionStage<>(completionStage, executor);
```

Comment:
Maybe this should be guarded:
```suggestion
        return completionStage instanceof InternalCompletionStage ? this : new InternalCompletionStage<>(completionStage, executor);
```

Reply:
Done, but included an additional guard that ensures that the executor is that expected.  The executor should never be anything other than Netty executor, but this still feels like a sensible precaution.